qpsolvers/solvers/nppro_.py

```diff
@@ -0,0 +1,207 @@
+        l_ = np.full(h.shape, -np.infty)
+        u_ = h
+    if A is not None and b is not None:
+        A_ = A if A_ is None else np.vstack([A_, A])
+        l_ = b if l_ is None else np.hstack([l_, b])
+        u_ = b if u_ is None else np.hstack([u_, b])
+    if lb is not None:
+        lb_ = lb
+    if ub is not None:
+        ub_ = ub
+
+    # Create solver object
+    solver = nppro.CreateSolver(n, m)
+
+    # Set options
+    solver.setOption_MaxIter(100);
+    solver.setOption_SkipPreprocessing(False);
+    solver.setOption_SkipPhaseOne(False);
+    solver.setOption_InfVal(1e16);
+    solver.setOption_HessianUpdates(True);
+
+    x0 = np.full(q.shape, 0)
+    if initvals is not None:
+        x0 = initvals
+
+    # Conversion to datatype supported by the solver's C++ interface
+    P = np.asarray(P, order='C', dtype=np.float64)
+    q = np.asarray(q, order='C', dtype=np.float64)
+    A_ = np.asarray(A_, order='C', dtype=np.float64)
+    l_ = np.asarray(l_, order='C', dtype=np.float64)
+    u_ = np.asarray(u_, order='C', dtype=np.float64)
+    lb_ = np.asarray(lb_, order='C', dtype=np.float64)
+    ub_ = np.asarray(ub_, order='C', dtype=np.float64)
+    x0 = np.asarray(x0, order='C', dtype=np.float64)
+
+    # Call solver
+    x, fval, exitflag, iter_ = solver.solve(P, q, A_, l_, u_, lb_, ub_, x0)
+
+    # Store solution
+    solution = Solution(problem)
+    if exitflag != 0 or np.isnan(x).any():
+        # The second condition typically handle positive semi-definite cases
+        # that are not catched by the solver yet
+        warnings.warn(f"NPPro exited with status {exitflag}")
+        return solution
+    solution.x = x
+    if G is not None:
+        solution.z = np.full(h.shape, 0) # not available yet
+    if A is not None:
+        solution.y = np.full(b.shape, 0) # not available yet
+    if lb is not None or ub is not None:
+        solution.z_box = np.full(q.shape, 0) # not available yet
+    solution.extras = {
+        "cost": fval,
+        "iter": iter_,
+    }
+    return solution
+
+
+def nppro_solve_qp(
+    P: np.ndarray,
+    q: np.ndarray,
+    G: Optional[np.ndarray] = None,
+    h: Optional[np.ndarray] = None,
+    A: Optional[np.ndarray] = None,
+    b: Optional[np.ndarray] = None,
+    lb: Optional[np.ndarray] = None,
+    ub: Optional[np.ndarray] = None,
+    initvals: Optional[np.ndarray] = None,
+    verbose: bool = False,
+    **kwargs,
+) -> Optional[np.ndarray]:
+    r"""Solve a quadratic program using NPPro.
+
+    The quadratic program is defined as:
+
+    .. math::
+
+        \begin{split}\begin{array}{ll}
+            \underset{x}{\mbox{minimize}} &
+                \frac{1}{2} x^T P x + q^T x \\
+            \mbox{subject to}
+                & G x \leq h                \\
+                & A x = b                   \\
+                & lb \leq x \leq ub
+        \end{array}\end{split}
+
+    It is solved using `NPPro`__.
+
+    Parameters
+    ----------
+    P :
+        Symmetric cost matrix.
```

Comment:
If you have assumptions on the cost matrix you can write them down here.

Reply:
assumption mentioned there is P: Positive definite cost matrix